src/audio/module_adapter/module_adapter.c

```diff
@@ -241,20 +241,24 @@ int module_adapter_prepare(struct comp_dev *dev)
 	}
 
 	/* allocate memory for input buffers */
-	mod->input_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+	if (mod->num_input_buffers) {
+		mod->input_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
 				     sizeof(*mod->input_buffers) * mod->num_input_buffers);
-	if (!mod->input_buffers) {
-		comp_err(dev, "module_adapter_prepare(): failed to allocate input buffers");
-		return -ENOMEM;
+		if (!mod->input_buffers) {
+			comp_err(dev, "module_adapter_prepare(): failed to allocate input buffers");
+			return -ENOMEM;
+		}
 	}
 
 	/* allocate memory for output buffers */
-	mod->output_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
-				      sizeof(*mod->output_buffers) * mod->num_output_buffers);
-	if (!mod->output_buffers) {
-		comp_err(dev, "module_adapter_prepare(): failed to allocate output buffers");
-		ret = -ENOMEM;
-		goto in_free;
+	if (mod->num_output_buffers) {
+		mod->output_buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					sizeof(*mod->output_buffers) * mod->num_output_buffers);
+		if (!mod->output_buffers) {
+			comp_err(dev, "module_adapter_prepare(): failed to allocate output buffers");
+			ret = -ENOMEM;
+			goto in_free;
```

Comment:
this isn't something that this patch is breaking, but clean up labels aren't correct here. Look at `goto in_free;` at line 321 / 325 below: it's leaking `mod->output_buffers`. The order of labels should be (choose whatever names you prefer):
```
free4:
	for (i = 0; i < mod->num_output_buffers; i++)
		rfree((__sparse_force void *)mod->output_buffers[i].data);
free3:
	for (i = 0; i < mod->num_input_buffers; i++)
		rfree((__sparse_force void *)mod->input_buffers[i].data);
free2:
	rfree(mod->output_buffers);
free1:
	rfree(mod->input_buffers);
	return ret;
```
Also if later more initialisation steps are added the respective clean up will just need to be added, the existing steps won't need to be changed

Reply:
Good point, thanks. Fixed.